libnavui-maps/src/main/java/com/mapbox/navigation/ui/maps/route/line/api/MapboxRouteLineView.kt

```diff
@@ -559,89 +478,386 @@ class MapboxRouteLineView(var options: MapboxRouteLineOptions) {
+        layerIds.forEach { layerId ->
+            style.getLayer(layerId)?.let {
+                (it as LineLayer).lineGradient(expression)
+            }
+        }
+    }
+
+    private fun findSourceForFeature(
+        sourceFeatureMap: Map<RouteLineSourceKey, RouteLineFeatureId>,
+        featureId: RouteLineFeatureId
+    ): RouteLineSourceKey? {
+        return sourceFeatureMap.filterValues { it == featureId }.mapNotNull { it.key }.firstOrNull()
+    }
+
+    private fun moveLayersUp(
+        style: Style,
+        sourceKey: RouteLineSourceKey,
+        sourceToLayerMap: Map<RouteLineSourceKey, Set<String>>
     ) {
-        if (expressionProvider != null) {
-            val gradientExpression = coroutineScope.async {
-                expressionProvider.generateExpression()
+        sourceToLayerMap[sourceKey]?.forEach {
+            style.moveStyleLayer(
+                it,
+                LayerPosition(null, RouteLayerConstants.TOP_LEVEL_ROUTE_LINE_LAYER_ID, null)
+            )
+        }
+    }
+
+    private fun getSourceKeyForPrimaryRoute(style: Style): Result<RouteLineSourceKey> {
+        return runCatching {
+            val upperRange = style.styleLayers.indexOf(
+                StyleObjectInfo(
+                    RouteLayerConstants.TOP_LEVEL_ROUTE_LINE_LAYER_ID,
+                    "background"
+                )
+            )
+            val lowerRange = style.styleLayers.indexOf(
+                StyleObjectInfo(
+                    RouteLayerConstants.BOTTOM_LEVEL_ROUTE_LINE_LAYER_ID,
+                    "background"
+                )
+            )
+
+            style.styleLayers.subList(lowerRange, upperRange).mapIndexed { index, styleObjectInfo ->
+                Pair(index, styleObjectInfo.id)
+            }.maxByOrNull { it.first }?.run {
+                when {
+                    this.second.contains(
+                        RouteLayerConstants.MAPBOX_ROUTE_LINE_LAYER_GROUP_PREFIX + "1",
+                        true
+                    ) -> {
+                        MapboxRouteLineUtils.layerGroup1SourceKey
+                    }
+                    this.second.contains(
+                        RouteLayerConstants.MAPBOX_ROUTE_LINE_LAYER_GROUP_PREFIX + "2",
+                        true
+                    ) -> {
+                        MapboxRouteLineUtils.layerGroup2SourceKey
+                    }
+                    this.second.contains(
+                        RouteLayerConstants.MAPBOX_ROUTE_LINE_LAYER_GROUP_PREFIX + "3",
+                        true
+                    ) -> {
+                        MapboxRouteLineUtils.layerGroup3SourceKey
+                    }
+                    else -> throw NoSuchElementException()
+                }
+            } ?: throw NoSuchElementException()
+        }
+    }
+
+    private fun getUpdateSourceCmd(
+        routeData: Pair<RouteLineSourceKey, RouteLineData>,
+        style: Style
+    ): suspend () -> Unit {
```

Comment:
```suggestion
    ): () -> Unit {
```
There's nothing suspending in the returned function, everything runs synchronously. This is also true for all the other "command" generated in this class.

Does that mean that we're potentially regressing because some of the updates or expression generation is done synchronously now?

Reply:
In early iterations of this work I was executing the expression generation in a way that made sense with the suspend functions so that the generation wasn't on the main thread and was async. However during my testing I realized that it was indeed synchronous even thought it wasn't on the main thread.  I fixed that issue without realizing I no longer needed the functions to be suspend functions.  Good catch.